File: admin/lib/models/support.dart
class SupportTicket {
  final String id;
  final String ticketNumber;
  final String? orderId;
  final String? customerId;
  final String customerEmail;
  final String subject;
  final String status;
  final String priority;
  final String? assignedAdminId;
  final DateTime createdAt;
  final DateTime updatedAt;

  const SupportTicket({
    required this.id,
    required this.ticketNumber,
    this.orderId,
    this.customerId,
    required this.customerEmail,
    required this.subject,
    required this.status,
    required this.priority,
    this.assignedAdminId,
    required this.createdAt,
    required this.updatedAt,
  });

  factory SupportTicket.fromJson(Map<String, dynamic> json) {
    return SupportTicket(
      id: json['id'] as String,
      ticketNumber: json['ticketNumber'] as String,
      orderId: json['orderId'] as String?,
      customerId: json['customerId'] as String?,
      customerEmail: json['customerEmail'] as String,
      subject: json['subject'] as String,
      status: json['status'] as String,
      priority: json['priority'] as String? ?? 'normal',
      assignedAdminId: json['assignedAdminId'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }
}

class TicketMessage {
  final String id;
  final String ticketId;
  final String senderType;
  final String? senderAdminId;
  final String? senderCustomerId;
  final String body;
  final bool isInternal;
  final DateTime createdAt;

  const TicketMessage({
    required this.id,
    required this.ticketId,
    required this.senderType,
    this.senderAdminId,
    this.senderCustomerId,
    required this.body,
    required this.isInternal,
    required this.createdAt,
  });

  factory TicketMessage.fromJson(Map<String, dynamic> json) {
    return TicketMessage(
      id: json['id'] as String,
      ticketId: json['ticketId'] as String,
      senderType: json['senderType'] as String,
      senderAdminId: json['senderAdminId'] as String?,
      senderCustomerId: json['senderCustomerId'] as String?,
      body: json['body'] as String,
      isInternal: json['isInternal'] as bool? ?? false,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

File: api/src/db/schema/support.ts
import { pgTable, uuid, text, timestamp, integer, boolean } from "drizzle-orm/pg-core";
import { customer } from "./customer.js";
import { order } from "./order.js";
import { shipment } from "./fulfillment.js";
import { adminUser } from "./admin.js";

// ---------------------------------------------------------------------------
// support_ticket
// ---------------------------------------------------------------------------

export const supportTicket = pgTable("support_ticket", {
  id: uuid().defaultRandom().primaryKey(),
  ticketNumber: text("ticket_number").notNull(),
  customerId: uuid("customer_id").references(() => customer.id),
  orderId: uuid("order_id").references(() => order.id),
  shipmentId: uuid("shipment_id").references(() => shipment.id),
  subject: text().notNull(),
  category: text().notNull(),
  priority: text().notNull().default("normal"),
  status: text().notNull().default("open"),
  source: text().notNull(),
  potentialDuplicate: boolean("potential_duplicate").notNull().default(false),
  linkedTicketId: uuid("linked_ticket_id"),
  duplicateDismissed: boolean("duplicate_dismissed").notNull().default(false),
  mergedIntoTicketId: uuid("merged_into_ticket_id"),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow(),
  resolvedAt: timestamp("resolved_at", { withTimezone: true }),
});

// ---------------------------------------------------------------------------
// support_ticket_message
// ---------------------------------------------------------------------------

export const supportTicketMessage = pgTable("support_ticket_message", {
  id: uuid().defaultRandom().primaryKey(),
  ticketId: uuid("ticket_id")
    .notNull()
    .references(() => supportTicket.id),
  authorType: text("author_type").notNull(),
  customerId: uuid("customer_id").references(() => customer.id),
  adminUserId: uuid("admin_user_id").references(() => adminUser.id),
  body: text().notNull(),
  isInternalNote: boolean("is_internal_note").notNull().default(false),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
});

// ---------------------------------------------------------------------------
// support_ticket_attachment
// ---------------------------------------------------------------------------

export const supportTicketAttachment = pgTable("support_ticket_attachment", {
  id: uuid().defaultRandom().primaryKey(),
  ticketId: uuid("ticket_id")
    .notNull()
    .references(() => supportTicket.id),
  messageId: uuid("message_id").references(() => supportTicketMessage.id),
  storageKey: text("storage_key").notNull(),
  fileName: text("file_name").notNull(),
  contentType: text("content_type").notNull(),
  sizeBytes: integer("size_bytes").notNull(),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
});

// ---------------------------------------------------------------------------
// support_ticket_status_history
// ---------------------------------------------------------------------------

export const supportTicketStatusHistory = pgTable("support_ticket_status_history", {
  id: uuid().defaultRandom().primaryKey(),
  ticketId: uuid("ticket_id")
    .notNull()
    .references(() => supportTicket.id),
  oldStatus: text("old_status").notNull(),
  newStatus: text("new_status").notNull(),
  actorAdminUserId: uuid("actor_admin_user_id").references(() => adminUser.id),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
});
